hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/supervisor/workload/SubscriptionAssignmentRegistry.java

```diff
@@ -101,6 +94,38 @@ public SubscriptionAssignmentView createSnapshot() {
         return new SubscriptionAssignmentView(snapshot);
     }
 
+    private List<SubscriptionAssignment> readExistingAssignments() {
+        return subscriptionsCache.listActiveSubscriptionNames().stream().flatMap(subscriptionName -> {
+            String path = pathSerializer.serialize(subscriptionName);
+            List<String> children = new ArrayList<>();
+            try {
+                children.addAll(curator.getChildren().forPath(path));
+            } catch (Exception e) {
+                // ignore - should be fixed by the cache
+            }
+            return children.stream().map(node -> new SubscriptionAssignment(node, subscriptionName));
+        }).collect(Collectors.toList());
+    }
+
+    private void onAssignmentAdded(SubscriptionAssignment assignment) {
+        assignments.add(assignment);
+        if (consumerNodeId.equals(assignment.getConsumerNodeId())) {
+            callbacks.forEach(callback ->
+                callback.onSubscriptionAssigned(assignment.getSubscriptionName())
+            );
+        }
+    }
+
+    private void onAssignmentRemoved(SubscriptionAssignment assignment) {
+        assignments.remove(assignment);
```

Comment:
i think we could check if this assignment was ever added, no need to do the job on all callbacks